processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java

```diff
@@ -407,6 +412,120 @@ private boolean handleDefinedMappings() {
             return errorOccurred;
         }
 
+        /**
+         * Iterates over all source parameter properties and maps them to current target.
+         * When a property name match occurs, the remainder will be checked for duplicates. Matches will be removed from
+         * the set of remaining target properties.
+         */
+        private void applySourcePropertiesToCurrentTarget( Mapping mapping ) {
```

Comment:
Isn't this method more or less the same as `applyPropertyNameBasedMapping`? Can we somehow combine it so we don't need to do a lot of manual things in here? I don't think that the methods that were added to support this method belong here really.

Reply:
applyPropertyNameBasedMapping was example for applySourcePropertiesToCurrentTarget (in other words I copied applyPropertyNameBasedMapping and changed it to do stuff with current target).
Main difference is that in `applyPropertyNameBasedMapping` it checks if there is property reader for method source parameter, while in `applySourcePropertiesToCurrentTarget` it uses source to get readers one level down (and use source from mapping not method). `applySourcePropertiesToCurrentTarget` also always maps to current current target object.
I would say 2 key difference are: 
source readers and SourceReference creation.
I guess mainly because `applySourcePropertiesToCurrentTarget` is called for mapping annotation, while `applyPropertyNameBasedMapping` is called for target properties without explicit mapping annotations.

Reply:
> Can we also have some more tests with second or third level nesting? I would like to see whether this works.

In CustomerMapper1 I have defined this:
```

    @Mapping( target = ".", source = "customer.item" )
    OrderItem map(OrderDTO order );

```

and code for it looks like this:
```

    public OrderItem map(OrderDTO order) {
        if (order == null) {
            return null;
        } else {
            OrderItem orderItem = new OrderItem();
            String id = this.orderCustomerItemId(order);
            if (id != null) {
                orderItem.setId(id);
            }

            int status = this.orderCustomerItemStatus(order);
            orderItem.setStatus(status);
            if (order.getOrderLines() != null) {
                Iterator var5 = order.getOrderLines().iterator();

                while(var5.hasNext()) {
                    OrderLineDTO orderLine = (OrderLineDTO)var5.next();
                    orderItem.addOrderLine(this.map(orderLine));
                }
            }

            orderItem.setCustomer(this.map(order.getCustomer()));
            return orderItem;
        }
    }

    private String orderCustomerItemId(OrderDTO orderDTO) {
        if (orderDTO == null) {
            return null;
        } else {
            CustomerDTO customer = orderDTO.getCustomer();
            if (customer == null) {
                return null;
            } else {
                ItemDTO item = customer.getItem();
                if (item == null) {
                    return null;
                } else {
                    String id = item.getId();
                    return id == null ? null : id;
                }
            }
        }
    }

    private int orderCustomerItemStatus(OrderDTO orderDTO) {
        if (orderDTO == null) {
            return 0;
        } else {
            CustomerDTO customer = orderDTO.getCustomer();
            if (customer == null) {
                return 0;
            } else {
                ItemDTO item = customer.getItem();
                if (item == null) {
                    return 0;
                } else {
                    int status = item.getStatus();
                    return status;
                }
            }
        }
    }


```

Reply:
What happens when the target is also nested? Perhaps there are already testcases covering this, but could it be that when stacking down, you ultimately hit `"."` or will it stop at an earlier level.

So:

`@Mapping( target = "store.shop", source = "store.shop.customer.item" )`

Will be stacked down to:

`@Mapping( target = "shop", source = "shop.customer.item" )`

and I guess it stops here (I guess we have testcase for this).

Reply:
when target is not `"."` it should be handled as before in `handleDefinedMapping`.